### DISCUSSION TITLE
Correlation using Rest API

### DISCUSSION THREAD

**Discussion by Rummy99** [elsa-workflows/elsa-core#1111] [Q&A] [answered]
Hi,

 

I am building workflow application using Elsa for the first time which involves many user interactions and approvals. User interface application and workflows will be running on different servers, so we are designing the workflows to be triggered only using Rest APIs. I have used dashboard to create the workflow. I am able to create a new workflow instance using HTTP Endpoint. I need to implement approval / reject use case for this instance.

1)      How to suspend the instance after the first submission and resume when approval / reject actions are performed? Approval / Rejection will be submitted via Rest APIs.

2)      How to Correlate the with the first submission using API endpoint?

**Comment by sfmskywalker** [comment, marked as answer]
Hi @Rummy99!

1. One activity that is suitable for that is the `Signal Received` activity, which will suspend the workflow and resume it once the appropriate signal is sent. See also [this guide](https://elsa-workflows.github.io/elsa-core/docs/next/guides/guides-document-approval) for an example.
2. When using the **HTTP Endpoint** activity as the workflow trigger, you can send the desired correlation ID along with the HTTP request either via the query string (e.g. `?correlation=foo`) or as an HTTP header (e.g. `X-Correlation-Id: foo`).

**Reply by Rummy99** [reply]
Hi Sipke,

Thank you for your reply.


After adding Correlation Id in request, new instance was not created. But I got another issue.

In my use case, the user should be able to save details as draft and later revisit the draft and do final submission for approval.  


My workflow starts with HTTP End Point, sets correlation ID, save details using a custom activity and send HTTP response. After response is sent I have added if Condition to check if the user action is Draft submission / Final submission.

1)      In case of Draft Submission, the workflow need to suspend and wait for the first HTTP end point action to be triggered again. The instance need to start gain from first activity in this case. I have suspended using a Signal Received. When HTTP Post request is submitted for next save action, I am getting 400 - BAD Request response.

2)      In case of Final Submission, I have added a Signal Received for Approval – Sending HTTP Get request with Signal name - Approve as URL is returning 400 - BAD Request response.


(I had included the correlation Id in the request header in above two cases)


Please guide me on what I have missed out here.

**Reply by sfmskywalker** [reply]
Can you take a look at the response body when you get the 400 code? It might contain a hint.

**Reply by Rummy99** [reply]
Hi Sipke,

My Signal_URL was wrong. It's working fine now. 

Thank a lot :)

**Comment by mdaizal** [comment]
Hi, I'm reviving this discussion as i'm currently having this problem. my approach is similiar to @Rummy99 . we have monolith old web apps so to integrate workflow engines and those apps can communicate with each other using elsa is via API. 
scenario is when a user submit a docs for approval (send to a httpendpoint) and the workflow started, an email will be sent to the approver with Accept and Reject signal url.(generated inside email following your guides.) all functioning smoothly if it happen in the same day(submit and approval). then i discovered a problem:
1. a user submit a doc for approval at 5th December 2022
2. the approver click the Accept/Reject button at 6th December 2022 - thus the problem arise - it got 400 bad request
here are the response that i got:

{"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"Bad Request","status":400,"traceId":"00-cfff181eafaffcb693c97bf4929a6ee6-efe39f132f53e441-00"}

I tried again to create the same problem, submit at 6th Dec, now today 7th Dec still got bad request. the problem persist.
I have no idea and stuck here.

btw, Elsa workflow rocks!

Please help. Thanks.

**Reply by mdaizal** [reply]
Forgot to add, my elsa is hosted in IIS, already set idle timeout=0 and always running. it has been running since 4th november without any issues.

**Reply by Rummy99** [reply]
Hi Aizal,

What is the persistence layer you are using. Looks like the correlation is not working after your app pool recycle.

**Reply by mdaizal** [reply]
by persistence did you means database? i'm sorry if i assume wrongly. i'm using sqlserver express 2017.
so iis app pool recycle kinda 'reset' my link? i'm quite new to iis too. it took me quite a while to understand and hosting Elsa in IIS.

**Reply by mdaizal** [reply]
i read again and again the answer by @sfmskywalker , does it means at the end of my signal url have to put `?correlation=mycorrelationid` ?

**Reply by mdaizal** [reply]
> Hi Aizal,
> 
> What is the persistence layer you are using. Looks like the correlation is not working after your app pool recycle.

I think you are correct. I manually recycle my app pool in IIS before clicking my signal link. and after that i got 400 bad request. now i set time interval for recyling to 0 (no auto recycle) and hopefully it will solve the problem. 

thank you.
